isofit/radiative_transfer/radiative_transfer.py

```diff
@@ -297,64 +293,61 @@ def drdn_dRT(
-            drdn_drfl = a / (1 - s_alb * rfl) ** 2
+        # atmospheric spherical albedo
+        s_alb = r["sphalb"]
 
-        elif self.glint_model:
-            L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
+        # direct and diffuse downward fluxes on the sun-to-surface path
+        # note: currently, E_down_dir comes scaled by the TOA solar zenith angle,
+        # thus, unscaling and rescaling by local solar zenith angle required
+        # to account for surface slope and aspect
+        E_down_dir, E_down_dif = self.get_E_down(x_RT, geom)
+        E_down_dir = E_down_dir / self.coszen * cos_i
 
-            E_dd = (self.solar_irr * self.coszen) / np.pi * r["t_down_dir"]
-            E_ds = (self.solar_irr * self.coszen) / np.pi * r["t_down_dif"]
-            E_d = E_dd + E_ds
-            L_sky = x_surface[-2] * E_dd + x_surface[-1] * E_ds
+        # including glint for water surfaces
+        if self.glint_model:
+            E_down_tot = E_down_dir + E_down_dif
+            L_sky = x_surface[-2] * E_down_dir + x_surface[-1] * E_down_dif
 
             rho_ls = 0.02  # fresnel reflectance factor (approx. 0.02 for nadir view)
-            glint = rho_ls * (L_sky / E_d)
-
-            drho_scaled_for_multiscattering_drfl = (
-                1.0 / (1 - r["sphalb"] * (rfl + glint)) ** 2
-            )
-            drdn_drfl = L_down_transmitted * drho_scaled_for_multiscattering_drfl
-
+            glint = rho_ls * (L_sky / E_down_tot)
         else:
-            L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
+            glint = np.zeros(rfl.shape)
 
-            # The reflected downwelling light is:
-            # L_down_transmitted * rfl / (1.0 - r['sphalb'] * rfl), or
-            # L_down_transmitted * rho_scaled_for_multiscattering
-            # This term is the derivative of rho_scaled_for_multiscattering
-            drho_scaled_for_multiscattering_drfl = 1.0 / (1 - r["sphalb"] * rfl) ** 2
+        # adjacency effects
+        bg = geom.bg_rfl if geom.bg_rfl is not None else rfl + glint
 
-            drdn_drfl = L_down_transmitted * drho_scaled_for_multiscattering_drfl
+        # K surface reflectance
+        drdn_drfl = (E_down_dir + E_down_dif) / (1.0 - s_alb * bg) * r["transm_up_dir"]
 
         drdn_dLs = r["transm_up_dir"] + r["transm_up_dif"]
+
         K_surface = (
             drdn_drfl[:, np.newaxis] * drfl_dsurface
             + drdn_dLs[:, np.newaxis] * dLs_dsurface
         )
 
+        if self.glint_model:
+            # K glint
+            drdn_dgdd = (
+                E_down_dir
+                * (r["transm_up_dir"] + r["transm_up_dif"])
+                / (1.0 - s_alb * bg)
+            )
+            drdn_dgdsf = (
+                E_down_dif
+                * (r["transm_up_dir"] + r["transm_up_dif"])
+                / (1.0 - s_alb * bg)
+            )
+
+            K_surface[:, -2] = drdn_dgdd
+            K_surface[:, -1] = drdn_dgdsf
```

Comment:
@pgbrodrick @davidraythompson I copied this from Regina's latest glint improvements.

Reply:
These need to be updated to some more general derivatives of additional surface parameters in a future PR. Resolving for now.